### ISSUE TITLE
Feature Request (max selectable )

### ISSUE THREAD

**Issue by smooJitter** [LauraBeatris/use-selected-items-hook#3]
it would be nice to add features to limit the number of items selected for example, 

a max (number) prop to capture the maximum allowed

an onMaxSwitchToNext (boolean) prop that says if the max is selected, the last selected is dropped and replaced by the currently selected item.  

Unfortunately, I am not familiar with the dependencies in this package so I can't so I am not able to submit a pull request.   Meanwhile, I can create another that wraps this hook.

Thanks for the contribution

**Comment by LauraBeatris** [comment]
Hey @smooJitter 

Thanks for the amazing feature request 

I'll definitely add it to the backlog!

**Comment by rodrigogama** [comment]
Hi @LauraBeatris,

I'd be glad to help you with this issue. How do you think we should address the solution? Should we put this logic inside the `reducer` or do some verification in `toggleSingleItem` and after dispatch an action?

@smooJitter
the prop `onMaxSwitchToNext`gives me the impression that it expects a function as a value instead of a boolean. What do you think about change it?

**Comment by LauraBeatris** [comment]
@rodrigogama Thanks 😄

> I'd be glad to help you with this issue. How do you think we should address the solution? Should we put this logic inside the reducer or do some verification in toggleSingleItem and after dispatch an action?

This logic should be applied for both actions - `toggleAllItems`  and `toggleSingleItem`. So it shouldn't be repeated on both handlers and also for both actions inside the reducer. 

The approach that I recommend is to create an external function to centralize the logic and then reuse it for both actions.

**Comment by gabrielnaoto** [comment]
> Hi @LauraBeatris,
> 
> I'd be glad to help you with this issue. How do you think we should address the solution? Should we put this logic inside the `reducer` or do some verification in `toggleSingleItem` and after dispatch an action?
> 
> @smooJitter
> the prop `onMaxSwitchToNext`gives me the impression that it expects a function as a value instead of a boolean. What do you think about change it?

I agree that this prop name, is not very intuitive. I think it could be something like `allowToggleNext`.

But on that subject. I think that this could be a problem applying this condition on a `toggleAllitems` action, because the `maxSelectedItems` could be smaller than the `initialItems.lenght`, in that case, which items would be selected?

Another point to consider, should the initialize reducer validate the initial selected items against the `maxSelectedItems` prop? To ensure that the state is always a valid one?

Another point is that, in order to allow toggling the last selected items I think we would need to have some type of queue to track the order on which items were selected.

And another thing I've noticed while testing this, is the toggle all items actions may note be behaving properly.

https://user-images.githubusercontent.com/13211186/106549638-2c1da100-64f0-11eb-8dfa-5a0df5d373ba.mov

When toggling all items, if there's one selected, should we just invert the selection, or instead, should we firstly select all items (if there are some not selected), and unselect all if all items are selected?

But, in general, I also think, that this could be a great feature, and I already see some use cases for that!

**Comment by LauraBeatris** [comment]
@gabrielnaoto Thanks for the comment! 

> I agree that this prop name, is not very intuitive. I think it could be something like allowToggleNext.

``allowToggleNext`` or ``shouldToggleNext`` are both intuitive property names and indicate booleans

> But on that subject. I think that this could be a problem applying this condition on a toggleAllitems action, because the maxSelectedItems could be smaller than the initialItems.lenght, in that case, which items would be selected?

If I understood the case, I'm thinking about this:
- `maxSelectedItems`: 4
- `initialItems`: 8

Only the first 4 items would be selected/toggled 

> Another point to consider, should the initialize reducer validate the initial selected items against the maxSelectedItems prop? To ensure that the state is always a valid one?

If a list of selected items is returned by the backend, it wouldn't be exact with the same length as the `maxSelectedItems` property. So I don't think that an error should be thrown while validating but then we just initialize the quantity defined in the `maxSelectedItems` property.

- `maxSelectedItems`: 4
- `initialItems`: 8
- `initialSelectedItems`: 5

In this case, only the first 4 items will be selected during initialization.

**Comment by LauraBeatris** [comment]
> Another point is that, in order to allow toggling the last selected items I think we would need to have some type of queue to track the order on which items were selected.

I don't if it would be too complex, so I'll also propose another approach. 

The logic is related to the lengths and indexes of the arrays. If we know that each item has a unique identifier we are able to track their position in the `initialItems` property. By knowing their initial position, it's possible to change the `isSelected` state of the last element to `false` and set the `isSelected` state of the next item to be selected as `true`.

`items` state
```
[{ id: 1, isSelected: true}, { id: 2, isSelected: false}, { id: 3, isSelected: true}]
```

Let's suppose I've selected the item with id `2`. Let's search for the index of that item and toggle its `isSelected` state, and then set the last selected item `isSelected` state to `false`.

```
[{ id: 1, isSelected: true}, { id: 2, isSelected: true}, { id: 3, isSelected: false}]
```


I think that would be better if all the logic related to max selectable items should be placed in a specific file and then the reducer module would be able to import and execute it. I'm afraid that the toggle logic will end up mixed and difficult to read. 

But feel free to use your judgment regarding what solution will be better in terms of readability and complexity.

**Comment by LauraBeatris** [comment]
> When toggling all items, if there's one selected, should we just invert the selection, or instead, should we firstly select all items (if there are some not selected), and unselect all if all items are selected?

Good point. I'm literally toggling the states so I didn't think about this case. Toggling `all` items can have a double meaning:
- Toggle all the items consistently: All `isSelected` states are updated together.
- Toggle all the items respecting each specific state: All `isSelected` states are updated according to each item state.

Each one is better? I can think of use cases for both 🤔

**Comment by LauraBeatris** [comment]
Alright! Thanks, everybody for sending suggestions. I'll start working on it ASAP 🔥

**Comment by gabrielnaoto** [comment]
> Each one is better? I can think of use cases for both 🤔

I was thinking about this and I think that `toggleAllItems` is correct, and it should behave as it is. 

But maybe, what we're missing here is the ability to `clearAllItems` and `selectAllItems`, which would clear or select, all items, respectively, regardless of their previous statuses.

But, then, we would need to have a prop that would tell if all items are selected or not `isAllSelected`, which would be available on the hook payload, and which would be what we would use to handle both cases.

I think this could be another issue just to discuss this. If you want I can open a new issue with this, and maybe a PR. But let me know what you think.
